Allow custom keywords that are not valid JS identifiers (fixes issue #389)

farrago opened the pull request:
**What issue does this pull request resolve?**
 
Custom keywords were required to be valid JS identifiers due to the way they were being used in the generated code.  This prevented the use of keywords with the `x-` prefix that is required by the [swagger](http://swagger.io) API definition spec.

See issue #389 for discussion.

**What changes did you make?**

This change fixes that by updating the only place that was using them directly as an identifier to use _brackets_ notation rather than _dot_ notation.  This brings the definition of `$schemaPath` in `definitions.def` in line with the way the keyword definition is accessed from `RULES.custom` at the top of `custom.jst`.

**Is there anything that requires more attention while reviewing?**

The validation of the keyword name has been changed to allow any name in this change, but this should be updated to either just remove the test or update it to a similar but more relaxed rule to the previous one 
before this change is included in the project.

**Tests**

This updates a number of the tests in `custom.spec.js` to use non-valid identifiers for the names of some of the custom keywords being tested (primarily by adding `x-` to the front of the name).

Test Plan:
- Run `npm test` and ensure all tests run and pass

<!--
Thank you for submitting a pull request to Ajv.

Before continuing, please read the guidelines:
https://github.com/epoberezkin/ajv/blob/master/CONTRIBUTING.md#pull-requests

If the pull request contains code please make sure there is an issue that we agreed to resolve (if it is a documentation improvement there is no need for an issue).

Please answer the questions below.
-->

farrago committed:
Allow custom keywords that are not valid JS identifiers

Custom keywords were required to be valid JS identifiers due to the
way they were being used in the generated code.  This prevented the use
of keywords with the `x-` prefix that is required by [swagger](http://swagger.io].

This change fixes that by updating the only place that was using them
directly as an identifier to use //brackets// notation rather than
//dot// notation.  This brings the definition of `$schemaPath` in
`definitions.def` in line with the way the rule definition is accessed
from `RULES.custom` at the top of `custom.jst`.

The validation of the keyword name has been changed to allow any name
in this change, but this should be updated to either just remove the
test or update it to a new rule before this change is included in
the project.

Finally, this also updates a number of the tests in `custom.spec.js` to
use non-valid identifiers for the names of some of the custom keywords
being tested (primarily by adding `x-` to the front of the name).

Test Plan:
- Run `npm test` and ensure all tests run and pass

epoberezkin commented on lib/keyword.js:
```diff
@@ -1,6 +1,6 @@
 
-var IDENTIFIER = /^[a-z_$][a-z0-9_$]*$/i;
+var IDENTIFIER = /^.*$/i;
 var customRuleCode = require('./dotjs/custom');
```
`/^[a-z_$][a-z0-9_$\-]*$/i`, as you've suggested

epoberezkin commented on lib/dot/definitions.def:
```diff
@@ -3,7 +3,7 @@
     var $lvl = it.level;
     var $dataLvl = it.dataLevel;
     var $schema = it.schema[$keyword];
-    var $schemaPath = it.schemaPath + '.' + $keyword;
```
it should be: `it.schemaPath + it.util.getProperty($keyword);`

This way it would still have direct property access when possible (which is in most cases and for all standard keywords)

farrago replied:
I did try this, but it gets copied directly to the generate code rather than being interpreted first.
i.e. you get a line in `lib\custom.js` that looks like:
```javascript
var $schemaPath = it.schemaPath + it.utils.getProperty($keyword);
```

when what we wanted is:
```javascript
var $schemaPath = it.schemaPath.foo; // or it.schemaPath['x-foo'] as appropriate
```

I couldn't work out how to make it insert the result of `getProperty()` rather than the literal value. So I just used hardcoded brackets notation directly as done in `custom.jst`:

```
23:   var {{=$definition}} = RULES.custom['{{=$keyword}}'].definition;
```

Can you suggest how to make this work (I'm not very familiar with dotjs)?

As an aside, is there a reason to favour one over the other?  I did do a quick search for performance comparisons, and while there was nothing conclusive it seemed that for fixed strings, brackets and dots are very similar in terms of performance.  Using variables with brackets notation is obviously slower as it can't optimise the lookup, but we are using a hardcoded string so it should be ok.

epoberezkin replied:
> i.e. you get a line in lib\custom.js that looks like:
> var $schemaPath = it.schemaPath + it.utils.getProperty($keyword);

If you mean the file lib\dotjs\custom.js then that's what it should be.
It's not a generated code of validation function, it's a code that generates code of validation function :)

epoberezkin replied:
The only question is if it works or not.

epoberezkin replied:
> I'm not very familiar with dotjs

It's a normal (and very simple) templating engine that allows JavaScript in it. Whatever is inside {{  }} is just code, so it is copied as is, while things in {{= }} get into the output.

It's a bit easier to manage template then it would be to manage string concatenation to generate code, that's it, but it's the same really.

farrago replied:
Ahh, I was trying to use it as a template parameter (i.e. wrap it in {{}}) which was making all sorts of wrong results.  But just copying it in directly does the right thing!  Well that was an hour wasted to my confusion!

I'll update with this and the other comments shortly.

epoberezkin reviewed (changes requested):
I will also need to double check that there are no other places where it can have direct property access.

farrago committed:
Improve implementation of allowing hyphens in keywords

This:
- Updates the README with the new rules (and recommendations)
- Updates `keyword.js` to implement the agreed change to the naming rules:
- - Now same as before but also allow hyphens as second and subsequent characters
- Update tests to test the new keyword naming rules
- Update `definitions.def` to use it.util.getProperty() to dynamically choose dot
  or bracket notation as appropriate rather than hardcode bracket notation.

Test Plan:
- Run `npm test` and ensure all tests run and pass
